src/Compiler/Facilities/DiagnosticOptions.fs

```diff
@@ -20,6 +20,8 @@ type FSharpDiagnosticOptions =
         WarnOn: int list
         WarnAsError: int list
         WarnAsWarn: int list
+        WarnScopesFeatureIsSupported: bool
+        mutable WarnScopeData: obj option
```

Comment:
Yes, I did not really like to add a mutable field, but here are my "excuses".
1. I believe that the information about the warn scopes belongs into `FSharpDiagnosticsOptions` as much as all the other fields. _All these fields_ are used only in one single place in the whole compiler, namely in `PhasedDiagnostics.AddSeverity`. Passing the information separately does not fit its purpose and would add an additional parameter to many functions.
2. This information needs to be updated during postparse for every file. This is in the nature of the data: the #nowarn/#warnon directives change the way warnings are emitted.
3. I modeled this in some way after the `BufferLocalStore` of `lexbuf`. Just a way to store data for a certain aspect for the current compiler phase.
4. `WarnScopeData` is accessed only via private functions `WarnScopes.getWarnScopeData` and `WarnScopes.setWarnScopeData` in the `WarnScopes` module.

I will look into alternatives and see if I find something acceptable.
(For fsc, we could probably go back to tcConfigB and update the diagnosticsOptions there. But this may not work well for the other use cases. And it would not be safer or nicer or more idiosyncratic than the mutable field.)